src/lexparse/lexer.go

```diff
@@ -3,73 +3,228 @@ package lexparse
 import (
   "fmt"
   "strings"
-)
-
-type tokenType int
-
-const (
-  SymbolT tokenType = iota
-
-  StringT tokenType = iota
-  CharT tokenType = iota
-  IntT tokenType = iota
-  FloatT tokenType = iota
-  BoolT tokenType = iota
-  NullT tokenType = iota
-
-  BinopT tokenType = iota // + - * / ^ % == != <= >=
-  SetterT tokenType = iota // = += -= *= /= ^= %=
-  PrefixT tokenType = iota // ! ~
-  Suffix tokenType = iota // $ #
-
-  LParT tokenType = iota
-  RParT tokenType = iota
-  LSquareBrT tokenType = iota
-  RSquareBrT tokenType = iota
-  LCurlyBrT tokenType = iota
-  RCurlyBrT tokenType = iota
-  LAngleT tokenType = iota
-  RAngleT tokenType = iota
-
-  DotT tokenType = iota
-  QMarkT tokenType = iota
-  EllipsisT tokenType = iota
-  CommaT tokenType = iota
-  SemicolonT tokenType = iota
-  ColonT tokenType = iota
+  "os"
+  "io/ioutil"
+  "unicode"
 )
 
 type Line struct {
   Num int
-  Value string
+  Value []rune
   Filename string
 }
 
 type Token struct {
-  Line Line
+  Line *Line
   Col int
   TType tokenType
-  Value string
+  Value []rune
 }
 
-func lexerPanic(token Token, message string) {
-  s := fmt.Sprintf(
-    "Teko Parser Error (%s:%d)\n%s\n%s%s\n%s",
-    token.Line.Filename,
-    token.Col,
-    token.Line.Value,
-    strings.Repeat(" ",token.Col),
-    strings.Repeat("^",len(token.Value)),
+func lexparsePanic(line *Line, col int, width int, message string) {
+  fmt.Printf(
+    "Teko Parser Error (%s:%d)\n%s\n%s%s\n%s\n",
+    line.Filename,
+    col,
+    string(line.Value),
+    strings.Repeat(" ",col),
+    strings.Repeat("^",width),
     message,
   )
-  panic(s)
+  os.Exit(0)
+}
+
+func tokenPanic(token Token, message string) {
+  lexparsePanic(token.Line, token.Col, len(token.Value), message)
+}
+
+type Lexer struct {
+  Line *Line
+  Col int
+  CurrentBlob []rune
+  InBlockComment bool
+  CurrentTType tokenType
+}
+
+func LexFile(filename string) []Token {
+  dat, err := ioutil.ReadFile(filename)
+  if err != nil {
+    panic(err)
+  }
+  contents := string(dat)
+
+  lines_raw := strings.Split(contents, "\n")
+  lines := make([]Line, len(lines_raw))
+
+  for i, s := range lines_raw {
+    lines[i] = Line{
+      Num: i,
+      Value: []rune(s),
+      Filename: filename}
+  }
+
+  tokens := []Token{}
+  lexer := Lexer{InBlockComment: false}
+
+  for _, line := range lines {
+    lexer.startline(&line)
+    tokens = append(tokens, lexer.grabTokens()...)
```

Comment:
The approach I've seen more often is to just treat newlines as special tokens. Don't know if it really makes a difference. One possible issue with this method is you can't have multiline tokens, e.g. multiline strings.

Reply:
:thinking: I have to think about this more. I think of Teko as generally totally oblivious to whitespace but you raise a good point - I hadn't made any plans about multiline strings but they're probably a good thing to have eventually.